### PR TITLE
[Loader] Changes to support ONNX graphs generated from tensorflow in NHWC format

### PR DESCRIPTION
**Pull request by dmoham1476**
*Description*:
GLOW compiler supports caffe2 and ONNX models in NCHW format. Using tf2onnx, ONNX graphs can be generated from tensorflow checkpoints. The input to the graph in this case is in the NHWC format.  The ONNX graphs generated from tensorflow models also have some operators like tf.Identity and tf.MatMul that is not currently supported by the GLOW compiler ONNX code. The pull request is for the changes to support this additional feature for googlenet_v1_slim.onnx and google_v4_slim.onnx generated from tensorflow pretrained models. 

*Testing*: Ran tests/images/run.sh, ran ninja test, all tests passed with "-DGLOW_WITH_OPENCL=OFF -DGLOW_WITH_CPU=OFF

*Documentation*: tf2onnx tool - https://github.com/onnx/tensorflow-onnx/tree/master/tf2onnx
Pretrained googlenet models - https://github.com/tensorflow/models/tree/master/research/slim#pre-trained-models

### CODE REVIEW

**Comment by dmoham1476** [comment]
Testing: Ran tests/images/run.sh, ran ninja test, all tests passed with "-DGLOW_WITH_OPENCL=OFF -DGLOW_WITH_CPU=OFF, Inference results using GLOW are consistent with tensorflow inference results.

File: tools/loader/ImageClassifier.cpp [hunk 1]
```diff
@@ -83,11 +83,18 @@ llvm::cl::opt<ImageNormalizationMode> imageMode(
 llvm::cl::alias imageModeA("i", llvm::cl::desc("Alias for -image_mode"),
                            llvm::cl::aliasopt(imageMode),
                            llvm::cl::cat(imageLoaderCat));
+
+
+llvm::cl::opt<std::string> loadimageFormat(
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: Instead of having this as a string, can you create an enum for it?

File: tools/loader/ImageClassifier.cpp [hunk 2]
```diff
@@ -83,11 +83,18 @@ llvm::cl::opt<ImageNormalizationMode> imageMode(
...
+    "image_format",
+    llvm::cl::desc("Specify which image format to use"));
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: All of our models up until now assume NCHW input -- can you add a default value here for NCHW?

File: tools/loader/ImageClassifier.cpp [hunk 3]
```diff
@@ -83,11 +83,18 @@ llvm::cl::opt<ImageNormalizationMode> imageMode(
...
+
 } // namespace
 
 /// Loads and normalizes all PNGs into a tensor in the NCHW 3x224x224 format.
 void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
-                             Tensor *result, ImageNormalizationMode normMode) {
+                             Tensor *result, ImageNormalizationMode normMode,
+                             const llvm::cl::opt<std::string> &imageformat) {
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: No need to pass the `imageFormat` around like this -- it's global in this file. You can just access it by name inside `loadImagesAndPreprocess()`

File: tools/loader/ImageClassifier.cpp [hunk 4]
```diff
@@ -135,11 +157,17 @@ int main(int argc, char **argv) {
...
   Tensor data;
-  loadImagesAndPreprocess(inputImageFilenames, &data, imageMode);
+
+  loadImagesAndPreprocess(inputImageFilenames, &data, imageMode, loadimageFormat);
 
   // All of our ONNX and Caffe2 models may use one of two names for input.
   constexpr char const *c2Inputs[2] = {"gpu_0/data", "data"};
-  constexpr char const *onnxInputs[2] = {"gpu_0/data_0", "data_0"};
+  
+  char const *onnxInputs[2] = {"gpu_0/data_0", "data_0"};
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: The only reason we have it set up like this is the models we have run so far all have two possible input names. Does the input model you are running actually have two different input variable name possibilities too? It might not be necessary for other models, and we perhaps should refactor this.

File: tools/loader/ImageClassifier.cpp [hunk 5]
```diff
@@ -99,9 +106,19 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
   std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filenames[0].c_str());
   const size_t numChannels = isGray ? 1 : 3;
 
+  if(imageformat != "NHWC") {
```
Comment (artemrakhov) [inline tools/loader/ImageClassifier.cpp]: It's more readable to write: `if (imageformat == "NCHW")`

File: tools/loader/ImageClassifier.cpp [hunk 6]
```diff
@@ -99,9 +106,19 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
...
   // N x C x H x W
-  result->reset(ElemKind::FloatTy,
+    result->reset(ElemKind::FloatTy,
                 {numImages, numChannels, imgHeight, imgWidth});
+  }
+  //N x H x W x C
```
Comment (artemrakhov) [inline tools/loader/ImageClassifier.cpp]: please move comment inside else branch, to match style of previous comment

File: tools/loader/ImageClassifier.cpp (hunk 1 above)
Comment (artemrakhov) [inline tools/loader/ImageClassifier.cpp]: We tend to use camelCase for variable names. Also, "format" is very general word. Let's use "layout"

loadimageFormat -> imageLayout

**Comment by rdzhabarov** [comment]
@dmoham1476
side note, can you please clean the description of the PR and remove

`
[Optional Fixes #issue]

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.
`

File: tools/loader/ImageClassifier.cpp [hunk 7]
```diff
@@ -121,7 +138,12 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
     for (unsigned z = 0; z < numChannels; z++) {
       for (unsigned y = 0; y < dims[1]; y++) {
         for (unsigned x = 0; x < dims[0]; x++) {
-          RH.at({n, numChannels - 1 - z, x, y}) = (imageH.at({x, y, z}));
+          if(imageformat != "NHWC") {
```
Comment (rdzhabarov) [inline tools/loader/ImageClassifier.cpp]: can you explicitly check against NCHW here.

File: tools/loader/ImageClassifier.cpp [hunk 8]
```diff
@@ -135,11 +157,17 @@ int main(int argc, char **argv) {
...
-  loadImagesAndPreprocess(inputImageFilenames, &data, imageMode);
+
+  loadImagesAndPreprocess(inputImageFilenames, &data, imageMode, loadimageFormat);
 
   // All of our ONNX and Caffe2 models may use one of two names for input.
   constexpr char const *c2Inputs[2] = {"gpu_0/data", "data"};
-  constexpr char const *onnxInputs[2] = {"gpu_0/data_0", "data_0"};
+  
+  char const *onnxInputs[2] = {"gpu_0/data_0", "data_0"};
+  if(loadimageFormat == "NHWC") {
```
Comment (rdzhabarov) [inline tools/loader/ImageClassifier.cpp]: I imagine there will be models with different input names, maybe worthwhile passing this (inputs) as a CLI parameter.
If the parameter is missing then use that default ```{"gpu_0/data_0", "data_0"};```

File: tools/loader/ImageClassifier.cpp [hunk 9]
```diff
@@ -83,11 +83,18 @@ llvm::cl::opt<ImageNormalizationMode> imageMode(
...
+
+
+llvm::cl::opt<std::string> loadimageFormat(
+    "image_format",
+    llvm::cl::desc("Specify which image format to use"));
+
 } // namespace
 
 /// Loads and normalizes all PNGs into a tensor in the NCHW 3x224x224 format.
 void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
```
Comment (nadavrot) [inline tools/loader/ImageClassifier.cpp]: Maybe a better technique would be to transpose the tensor outside of this method. It would simplify things.

File: tools/loader/ImageClassifier.cpp (hunk 6 above)
Comment (opti-mix) [reply tools/loader/ImageClassifier.cpp]: Please do not assume that anything that is not NCHW is automatically NHWC. We may add more formats in the future. So, e.g. assert inside else that it is NHWC

File: tools/loader/ImageClassifier.cpp (hunk 7 above)
Comment (opti-mix) [reply tools/loader/ImageClassifier.cpp]: Please do not use if-else to distinguish between formats. We may have more than 2 over time. Explicitly check for the formats you want to process and assert if you see an unexpected format.

File: include/glow/Importer/CommonOperatorLoader.h [hunk 10]
```diff
@@ -319,11 +319,9 @@ class CommonOperatorLoader : public ProtobufLoader {
     addNodeAsOutput(op, node);
   }
 
-  void loadDropout(const OpType &op, ArgumentDictionaryTy &dict) {
+  //void loadDropout(const OpType &op, ArgumentDictionaryTy &dict) {
```
[the remark this answers is not in the document]
Comment (dmoham1476) [reply include/glow/Importer/CommonOperatorLoader.h]: Will do

File: lib/Importer/ONNX.cpp [hunk 11]
```diff
@@ -442,6 +442,16 @@ bool ONNXModelLoader::loadOperator(const ONNX_NAMESPACE::NodeProto &op) {
     loadTranspose(op, dict, "perm");
     return true;
   }
+  if (typeName == "MatMul") {
+    auto LHS = getNodeValueOrCreateVariableByName(op.input(0));
+    auto RHS = getNodeValueOrCreateVariableByName(op.input(1));
+
+    Node *node = nullptr;
```
[the remark this answers is not in the document]
Comment (dmoham1476) [reply lib/Importer/ONNX.cpp]: Will do

File: tools/loader/ImageClassifier.cpp (hunk 1 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Hi Jordon, 

I made the following changes - 
enum class ImageLayout {
  /// N x C x H x W 
  NCHW,
  /// N x H x W x C
  NHWC,
};
ImageLayout strToImageLayout(const std::string &str) {
  return llvm::StringSwitch<ImageLayout>(str)
      .Case("nhwc", ImageLayout::NHWC)
      .Case("nchw", ImageLayout::NCHW);
  GLOW_ASSERT(false && "Unknown image layout");
}
llvm::cl::opt<ImageLayout> imageLayout(
    "image_layout", llvm::cl::desc("Specify which image layout to use"), 
    llvm::cl::Optional,
    llvm::cl::values(
        clEnumValN(ImageLayout::NCHW, "nchw",
                   "Use NCHW image layout"),
        clEnumValN(ImageLayout::NHWC, "nhwc",
                   "Use NHWC image layout")),
    llvm::cl::init(ImageLayout::NCHW));

I can see that it is being passed on the command line correctly, I am using the CLI passed in this function - 
void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
                             Tensor *result, ImageNormalizationMode normMode, ImageLayout layout) 

switch (layout) {
  case ImageLayout::NCHW:
    result->reset(ElemKind::FloatTy,
                {numImages, numChannels, imgHeight, imgWidth});
  case ImageLayout::NHWC:
    result->reset(ElemKind::FloatTy,
                {numImages, imgHeight, imgWidth, numChannels});
  default:
    GLOW_ASSERT(false && "Invalid Image Layout");
  }

I can see that the command line parameter takes in 0 for nchw and 1 for nhwc, when I print it inside this function. However, it is not being matched correctly in the switch/case (although the enum also takes int values). I am puzzled that I am missing some link here. I referred to the imageMode code and I cannot figure where the strToImageNormalizationMode function is being called or if it is needed.

File: tools/loader/ImageClassifier.cpp (hunk 2 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Indeed the current implementation is if you don't explicitly specify the imageformat, it does whatever is was earlier (NCHW). However this feedback has been given by most of you and hence I am working on making this change.

File: tools/loader/ImageClassifier.cpp (hunk 4 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: I am not sure for now, the models I have currently evaluated have only one name. So I would like to keep it this way for now, as I validate more models, I will keep this in mind.

File: tools/loader/ImageClassifier.cpp (hunk 5 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Will do

File: tools/loader/ImageClassifier.cpp (hunk 6 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Will do

File: tools/loader/ImageClassifier.cpp (hunk 1 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Will do

File: tools/loader/ImageClassifier.cpp (hunk 7 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Will do

File: tools/loader/ImageClassifier.cpp (hunk 8 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Can this be a separate change list? I do not want to combine this as part of this pull request. Is that ok? I believe there can be more underlying changes for this.

Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]: Sure, feel free to do this on separate PR. Smaller PRs are easier to review :)

File: tools/loader/ImageClassifier.cpp (hunk 1 above)
Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]: changed formatting, you can use "```" and cpp to format correctly.
in your switch you are missing breaks. Not sure if it's the only problem though.

Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: it worked, thanks :)

File: tools/loader/ImageClassifier.cpp (hunk 3 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Will do

**Comment by tlepley-cadence** [comment]
@dmoham1476 Please squash our commits when you will be done with addressing feedbacks. There are already 8 commits in your PR, and I my assumption is that one single commit would be sufficient.

File: tools/loader/ImageClassifier.cpp [hunk 12]
```diff
@@ -139,7 +178,13 @@ int main(int argc, char **argv) {
 
   // All of our ONNX and Caffe2 models may use one of two names for input.
   constexpr char const *c2Inputs[2] = {"gpu_0/data", "data"};
-  constexpr char const *onnxInputs[2] = {"gpu_0/data_0", "data_0"};
+  
+  // TODO add a CLI paramater to take in the input name of the graph.
```
[the remark this answers is not in the document]
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: So I wait for your changes to be merged in the main branch and then pull from that?

File: include/glow/Importer/CommonOperatorLoader.h [hunk 13]
```diff
@@ -437,9 +434,18 @@ class CommonOperatorLoader : public ProtobufLoader {
       return true;
     }
     if (typeName == "Dropout") {
-      loadDropout(op, dict);
+    // Save the identity operation. Note the second output (mask) for Dropout in
```
Comment (rdzhabarov) [inline include/glow/Importer/CommonOperatorLoader.h]: also, make sure to run clang-format.

File: tools/loader/ImageClassifier.cpp [hunk 14]
```diff
@@ -99,9 +119,19 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
...
+  switch (imageLayout) {
+  case ImageLayout::NCHW:
+    result->reset(ElemKind::FloatTy,
                 {numImages, numChannels, imgHeight, imgWidth});
+    break;
+  case ImageLayout::NHWC:
+    result->reset(ElemKind::FloatTy,
+                {numImages, imgHeight, imgWidth, numChannels});
+    break;
+  default:
```
Comment (rdzhabarov) [inline tools/loader/ImageClassifier.cpp]: remove `default` here, a compiler will ensure that case covers all values from `enum class`

File: tools/loader/ImageClassifier.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]: yup, just pull remote master changes into your local master and rebase this branch on top of that.

[the remark this answers is not in the document]
Comment (jfix71) [reply tools/loader/ImageClassifier.cpp]: Correct -- when you merge/rebase you will have conflicts here. Then just remove all of your changes here related to the input names, and then you can provide the name of the input to your model on the command line using `-model_input_name` (or whatever we settle on the name for this).

File: tools/loader/ImageClassifier.cpp [hunk 15]
```diff
@@ -99,9 +119,19 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
   std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filenames[0].c_str());
   const size_t numChannels = isGray ? 1 : 3;
 
-  // N x C x H x W
-  result->reset(ElemKind::FloatTy,
+  switch (imageLayout) {
```
[the remark this answers is not in the document]
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Hi I tried to remove the 2 switch/cases in the loadImagesAndPreprocess (keeping it as NCHW) and then I did this in the main() function, the build was ok, but when I run inference, I see that the inference result is wrong for NHWC. 

Tensor data;
  loadImagesAndPreprocess(inputImageFilenames, &data, imageMode)

if(imageLayout == ImageLayout::NHWC) {
       data.transpose(&data, NCHW2NHWC);
  }

File: include/glow/Importer/CommonOperatorLoader.h (hunk 13 above)
Comment (dmoham1476) [reply include/glow/Importer/CommonOperatorLoader.h]: yes, i had some issues wrt to the version of clang-format I was using, I will rerun with v6.0

File: tools/loader/ImageClassifier.cpp (hunk 14 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: So I had not added the default as there was not a case where I could hit the default condition (the llvminit took care of this). However somebody else commented that I should add the llvm_unreachable as the default: So should I remove this or not?

Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]: What was the motivation for llvm_unreachable?

The difference in behavior happens if you add a new enum option to the `enum class`.

On the one hand, If you do not have `default` option in the `switch` case you'll end up getting a compilation error that not all values are tested against in the `switch`. So, you'll have to fix this piece of code and add handling of the new enum value.

On the other hand, having `default` in place under the same circumstances will cause release build to ignore it and run with uninitialized result tensor (produce weird results). Debug build will fail with the assertion.

It feels that adding new layout should go along with the handling of it. Hence I'd prefer the first option.

File: tools/loader/ImageClassifier.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Sorry typo in the above line. I see that data in the main function is being passed into caffemodelloader, onnxmodelloader and also the run.inference function. The transpose function in tensor.h has a source and dest. I was thinking of adding the transpose like this -

if (imageLayout == ImageLayout::NHWC) {
    Tensor dataNHWC;
    dataNCHW.transpose(&dataNHWC, NCHW2NHWC)
  }

Then if need if/else switches to pass either dataNCHW or dataNHWC into the ONNX model loader and then the run.inference, so a lot of if/elses again in the main code. So I think the way it is now is not bad either. Please let me know what this saves us?.

**Comment by rdzhabarov** [comment]
@dmoham1476 are you blocked on anything?

File: tools/loader/ImageClassifier.cpp (hunk 14 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: Thank you, I will remove it

**Comment by dmoham1476** [comment]
@rdzhabarov I merged with jfix's code changes and validated my changes again and it works like before. Kindly approve the PR and then I can send a copy of the 2 models and image files so I can get that added to run.sh

File: include/glow/Importer/CommonOperatorLoader.h [hunk 16]
```diff
@@ -437,7 +434,15 @@ class CommonOperatorLoader : public ProtobufLoader {
       return true;
     }
     if (typeName == "Dropout") {
-      loadDropout(op, dict);
+      // Save the identity operation. Note the second output (mask) for Dropout
+      // in
```
Comment (rdzhabarov) [inline include/glow/Importer/CommonOperatorLoader.h]: can you please fix the flow here?

File: lib/Backends/CPU/LLVMIRGen.cpp [hunk 17]
```diff
@@ -1863,7 +1863,7 @@ void LLVMIRGen::generateLLVMIRForInstr(llvm::IRBuilder<> &builder,
     break;
   }
 
-    // Alloc and Dealloc instructions are handled by the memory allocator.
+  // Alloc and Dealloc instructions are handled by the memory allocator.
```
Comment (rdzhabarov) [inline lib/Backends/CPU/LLVMIRGen.cpp]: this seems reasonable, but could you extract that into separate commit? Just make sure that your changes are clang formatted.

File: lib/Backends/CPU/libjit/libjit.cpp [hunk 18]
```diff
@@ -458,8 +458,8 @@ void libjit_transpose_generic(const T *inW, T *outW, const size_t *idim,
 
   if (numDims == 2) {
     // Process the tiles in the matrix:
-    for (size_t sx = 0; sx < odim[0]; sx+=tileSize) {
-      for (size_t sy = 0; sy < odim[1]; sy+=tileSize) {
+    for (size_t sx = 0; sx < odim[0]; sx += tileSize) {
```
Comment (rdzhabarov) [inline lib/Backends/CPU/libjit/libjit.cpp]: same here.

File: tools/loader/ImageClassifier.cpp [hunk 19]
```diff
@@ -105,9 +124,17 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
   std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filenames[0].c_str());
   const size_t numChannels = isGray ? 1 : 3;
 
-  // N x C x H x W
-  result->reset(ElemKind::FloatTy,
-                {numImages, numChannels, imgHeight, imgWidth});
+  switch (imageLayout) {
```
Comment (rdzhabarov) [inline tools/loader/ImageClassifier.cpp]: Could you try the following:
* Remove this switch statement and switch statement in the loop below:
```
for (unsigned z = 0; z < numChannels; z++) {
      for (unsigned y = 0; y < dims[1]; y++) {
        for (unsigned x = 0; x < dims[0]; x++) {
          RH.at({n, numChannels - 1 - z, x, y}) = (imageH.at({x, y, z}));
======         
 switch (imageLayout) {
          case ImageLayout::NCHW:
            RH.at({n, numChannels - 1 - z, x, y}) = (imageH.at({x, y, z}));
            break;
          case ImageLayout::NHWC:
            RH.at({n, x, y, numChannels - 1 - z}) = (imageH.at({x, y, z}));
            break;
          }
=====
```

* Right after loadImagesAndPreprocess call, add another switch statement which does transposition when imageLayout == ImageLayout::NHWC and do nothing for ImageLayout::NCHW.

Would this work?

File: utils/format.sh [hunk 20]
```diff
@@ -13,4 +13,4 @@ fi
 find lib tests/unittests/ tools/ include examples \
      -name \*.h -print0 \
      -o -name \*.cpp -print0 \
-    | xargs -0 -P8 -n1 clang-format -i
+    | xargs -0 -P8 -n1 /usr/bin/clang-format-6.0 -i
```
Comment (rdzhabarov) [inline utils/format.sh]: that should not be in this PR (stick to a certain version of clang-format is a good idea though).

Comment (nadavrot) [reply utils/format.sh]: I don't have clang-format-6.0 on my computer.

File: lib/Backends/CPU/libjit/libjit.cpp (hunk 18 above)
Comment (dmoham1476) [reply lib/Backends/CPU/libjit/libjit.cpp]: this was automatically done when i ran clang-format on my repo.

File: utils/format.sh (hunk 20 above)
Comment (dmoham1476) [reply utils/format.sh]: sorry about this

File: lib/Backends/CPU/LLVMIRGen.cpp (hunk 17 above)
Comment (rdzhabarov) [reply lib/Backends/CPU/LLVMIRGen.cpp]: I understand that, but these files do not belong to the PR changes.

File: tools/loader/ImageClassifier.cpp (hunk 19 above)
Comment (dmoham1476) [reply tools/loader/ImageClassifier.cpp]: I tried this, I also commented about this to jfix's comment earlier today. the transpose function has a source/dest. So I created another Tensor dataNHWC and tried dataNCHW.transpose(&dataNHWC,NCHW2NHWC) with the NHWC switch after loadimageAndpreprocess call in the main(). But then data is used in onnxmodelloader and in run.inference, where I will have to pass either data or dataNHWC. So that is again 2 more if/else's

File: utils/format.sh (hunk 20 above)
Comment (dmoham1476) [reply utils/format.sh]: sorry about this

File: tools/loader/ImageClassifier.cpp (hunk 19 above)
Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]:
```
loadImagesAndPreprocess(inputImageFilenames, &data, imageMode);
switch (imageLayout) {
  case ImageLayout::NCHW:
    break;
  case ImageLayout::NHWC: {
    Tensor result;
    data.transpose(&result, NCHWTONHWC);
    data.assign(&result);
    break; 
  }
}
```

Do you need to add any more switches?

Comment (rdzhabarov) [reply tools/loader/ImageClassifier.cpp]: dataNHWC - sounds like a good idea, and also dataNCHW maybe.

**Comment by rdzhabarov** [comment]
@dmoham1476 please address @jfix71 comment before merging

File: tools/loader/ImageClassifier.cpp [hunk 21]
```diff
@@ -84,11 +89,25 @@ llvm::cl::alias imageModeA("i", llvm::cl::desc("Alias for -image_mode"),
...
+llvm::cl::alias imageLayoutA("l", llvm::cl::desc("Alias for -image_layout"),
+                             llvm::cl::aliasopt(imageLayout),
+                             llvm::cl::cat(imageLoaderCat));
+
 llvm::cl::opt<std::string> modelInputName(
     "model_input_name",
     llvm::cl::desc("The name of the variable for the model's input image."),
     llvm::cl::value_desc("string_name"), llvm::cl::Required,
     llvm::cl::cat(imageLoaderCat));
+
```
Comment (artemrakhov) [inline tools/loader/ImageClassifier.cpp]: nit: redundant empty line?

File: tools/loader/ImageClassifier.cpp [hunk 22]
```diff
@@ -108,6 +127,7 @@ void loadImagesAndPreprocess(const llvm::cl::list<std::string> &filenames,
   // N x C x H x W
   result->reset(ElemKind::FloatTy,
                 {numImages, numChannels, imgHeight, imgWidth});
+
```
Comment (artemrakhov) [inline tools/loader/ImageClassifier.cpp]: not related to the PR as well?

**Comment by artemrakhov** [comment]
Also, could you please shorten your commit message? Insert 2 linebreaks after `Changes to support onnx graphs generated from tensorflow in NHWC.`

**Comment by rdzhabarov** [comment]
Thanks @dmoham1476 for your patience and quickly addressing all the feedback on this PR.
